### PR TITLE
Peers Stack

### PR DESCRIPTION
**Pull request by breerly**
Collapsed for reviewability.

cc @willhug

### CODE REVIEW

File: peer/agent.go [hunk 1]
```diff
@@ -18,24 +18,22 @@
...
+	NotifyStatusChanged(Identifier)
 }
 
-// Agent manages Peers across different PeerSubscribers.  A PeerSubscriber will request a Peer for a specific
+// Agent manages Peers across different Subscribers.  A Subscriber will request a Peer for a specific
 // PeerIdentifier and the Agent has the ability to create a new Peer or return an existing one.
 type Agent interface {
-	PeerSubscriber
+	// Get or create a Peer for the Subscriber
+	RetainPeer(Identifier, Subscriber) (Peer, error)
```
Comment (breerly) [inline peer/agent.go]: The naming on this seems funny... why not just call it `GetPeer` where the semantics are get or create?

File: transport/http/agent_test.go [hunk 2]
```diff
@@ -4,374 +4,256 @@ import (
...
+				RetainAction{InputIdentifierID: "i1", InputSubscriberID: "s1", ExpectedPeerID: "i1"},
+				RetainAction{InputIdentifierID: "i1", InputSubscriberID: "s2", ExpectedPeerID: "i1"},
+				RetainAction{InputIdentifierID: "i1", InputSubscriberID: "s3", ExpectedPeerID: "i1"},
+				ReleaseAction{InputIdentifierID: "i1", InputSubscriberID: "s1"},
+				ReleaseAction{InputIdentifierID: "i1", InputSubscriberID: "s2"},
+				ReleaseAction{InputIdentifierID: "i1", InputSubscriberID: "s3"},
+			},
+		},
+		{
+			msg:         "no retains one release",
```
Comment (breerly) [inline transport/http/agent_test.go]: These messages seem to only describe how the world was setup, they don't say anything about the expected behavior. I would expect something more like:

```
err no reference when release called on empty peer list
```

The same seems to go for alot of these messages.

File: transport/http/agent_test.go [hunk 3]
```diff
@@ -4,374 +4,256 @@ import (
...
+				{ID: "s2"},
+				{ID: "s3"},
+				{ID: "s4"},
+				{ID: "s5rnd"},
+				{ID: "s6rnd"},
+				{ID: "s7rnd"},
+			},
+			actions: []AgentAction{
+				// Retains/Releases of i1 (Retain/Release the random peers at the end)
+				RetainAction{InputIdentifierID: "i1", InputSubscriberID: "s5rnd", ExpectedPeerID: "i1"},
```
Comment (breerly) [inline transport/http/agent_test.go]: What does `InputIdentifierID` mean? Why not just `InputID`?

File: transport/http/outbound.go [hunk 4]
```diff
@@ -86,7 +86,7 @@ func parseURL(urlStr string) (*url.URL, string) {
 // for getting potential downstream hosts.
 // PeerList.ChoosePeer MUST return *hostport.Peer objects.
 // PeerList.Start MUST be called before Outbound.Start
-func NewPeerListOutbound(peerList transport.PeerList, urlTemplate *url.URL) *Outbound {
+func NewPeerListOutbound(peerList peer.List, urlTemplate *url.URL) *Outbound {
```
Comment (breerly) [inline transport/http/outbound.go]: How does this play into oneway? Should this be a `NewPeerListUnaryOutbound`?

File: transport/http/outbound.go [hunk 5]
```diff
@@ -197,33 +162,44 @@ func (o *Outbound) CallOneway(ctx context.Context, treq *transport.Request) (tra
 		// panic because there's no recovery from this
 		panic(errOutboundNotStarted)
```
Comment (breerly) [inline transport/http/outbound.go]: Should we be panicking like this during runtime?

Comment (willhug) [inline transport/http/outbound.go]: probably not...

File: transport/http/outbound.go (hunk 4 above)
Comment (willhug) [inline transport/http/outbound.go]: Nope, oneway implements a CallOneway (vs Call) function so the http.PeerListOutbound is compatible with both.

File: transport/http/agent_test.go (hunk 3 above)
Comment (willhug) [inline transport/http/agent_test.go]: but VERBOSENESS!

I can change it (probably in a followup diff)

File: transport/http/agent_test.go (hunk 2 above)
Comment (willhug) [inline transport/http/agent_test.go]: Makes sense, I can do a runthrough to change.

**Pull request closed without merging**